src/java.base/share/classes/jdk/internal/foreign/ConfinedSession.java

```diff
@@ -81,10 +81,12 @@ public void release0() {
 
     void justClose() {
         checkValidState();
-        if (state == 0 || state - ((int)ASYNC_RELEASE_COUNT.getVolatile(this)) == 0) {
+        int asyncCount = (int)ASYNC_RELEASE_COUNT.getVolatile(this);
+        if ((state == 0 && asyncCount == 0)
+                || ((state - asyncCount) == 0)) {
             state = CLOSED;
         } else {
-            throw alreadyAcquired(state);
+            throw alreadyAcquired(state - asyncCount);
```

Comment:
I figured the exception message should take asyncCount into account as well. either both are 0, or state  - asynCount should be 0 for close to succeed. So, if we get here, we should always get a non-zero value in the exception message.